Feature/holo 1480 lazy mint

Yashiru opened the pull request:
## Describe Changes

- Add lazy mint feature (onlyOwner function to set the encrypted data for each batch)
- Add `canLazyMint` but we need to define if it is usefull or not and if yes => what mechanic to include in it
- Add tokenURI feature 
- Started to work on the test structure (Fixtures and constants)
- Started coding a first test to test the complete workflow in order to confirm it before writing it into the notion (at the very beginning, 3 lines of code)
- Add contractURI functionality

Yashiru committed:
feat: add LazyMint (WIP)

Yashiru committed:
feat: added baseUri support with lazyMint and delayed reveal (WIP)

Yashiru committed:
feat: prettier

Yashiru committed:
Revert "feat: prettier" customERC721 deletion

This reverts some of the changes in the commit 593d94027df5ada24662a83ed27570d179319406.

Yashiru committed:
fix: lazyMint usage

alexanderattar commented on src/HolographGenesis.sol:
```diff
@@ -246,11 +246,7 @@ contract HolographGenesis {
    * @param approve Approval status in the message
    * @return bytes32 Ethereum signed message hash
    */
-  function getMessageHash(
```
Uh oh @Yashiru I think we need to sync our prettier settings. These files shouldn't have this code updated like this

Yashiru replied:
yep you are right

Maybe we can use forge, it will probably not change any contract bytecode

alexanderattar commented on src/token/CustomERC721.sol:
```diff
@@ -430,13 +456,14 @@ contract CustomERC721 is NonReentrant, ERC721H, BatchMintMetadata, DelayedReveal
   ) external onlyOwner canMintTokens(recipients.length) returns (uint256) {
     unchecked {
-      for (uint256 i = 0; i < recipients.length; i++) {
+      for (uint256 i = 0; i != recipients.length; i++) {
```
Is this just a gas optimization?

Yashiru replied:
yes exactly

alexanderattar replied:
cool cool

alexanderattar commented on src/token/CustomERC721.sol:
```diff
@@ -23,14 +23,17 @@ import {SalesConfiguration} from "../drops/struct/SalesConfiguration.sol";
  */
-contract CustomERC721 is NonReentrant, ERC721H, BatchMintMetadata, DelayedReveal, ICustomERC721 {
+contract CustomERC721 is NonReentrant, ERC721H, LazyMint, DelayedReveal, ICustomERC721 {
+  using Strings for uint256;
```
Do we need this library?

Yashiru replied:
not anymore, I used it before my last commit

alexanderattar replied:
okay cool got it

alexanderattar commented on src/extension/LazyMint.sol:
```diff
@@ -0,0 +1,56 @@
+  error LazyMintInvalidAmount();
+
+  /// @notice The tokenId assigned to the next new NFT to be lazy minted.
+  uint256 internal nextTokenIdToLazyMint;
```
Okay so under the hood this keeps track of the sequence of tokens that have the next encrypted tokenURI. 

i.e. `nextTokenIdToLazyMint` - `amount`

so for example the next tokenId is 101 and the amount is 100, the next batch would be from 101-201

Yashiru replied:
yes exaclty

Yashiru committed:
feat: added tests

alexanderattar commented on src/token/CustomERC721.sol:
```diff
@@ -432,7 +432,7 @@ contract CustomERC721 is NonReentrant, ERC721H, LazyMint, DelayedReveal, ICustom
     string calldata _baseURIForTokens,
     bytes calldata _data
-  ) public override onlyOwner returns (uint256 batchId) {
+  ) public override returns (uint256 batchId) {
```
We should not remove `onlyOwner` from this. It needs to only be callable by the deployer of the contract

Yashiru replied:
Yes of course, It is mandatory, I don't know why it is removed

Yashiru replied:
Oh I know !

I just copy/past the version from thirdweb exemple to confirm that my function was the same, but their version has not the `onlyOwner` 👍🏻

alexanderattar commented on test/foundry/utils/Constants.sol:
```diff
@@ -67,4 +67,37 @@ library Constants {
   function getDropsEventConfig() internal pure returns (uint256) {
     return 0x0000000000000000000000000000000000000000000000000000000000040000;
   }
+
```
Seeing that this is growing, I think it would be a good idea to create a new directory under `foundry` called `constants` and then create a new file for contract specific constants. We can decide on a naming convention, but maybe something like `CustomERC721.Constants.sol`. `Core.Constants.sol`, etc

Yashiru committed:
feat: full flow tests

Yashiru committed:
test: CustomERC721 purchase test

alexanderattar committed:
Add support for contractURI

alexanderattar committed:
Update _canLazyMint to be only owner and add placeholder for claim

Yashiru committed:
test: cleanup and refactor

Yashiru committed:
test: full flow test (WIP)

- added purchase setup
- lazy mint by the owner using encrypted data
- actual purchase
- reveal tokenURI

Yashiru committed:
Merge branch 'feature/HOLO-1480-delay-reveal-countdown-erc721' into feature/HOLO-1480-lazy-mint

Yashiru committed:
feat: improved lazyMint script and fix provenanceHash

Yashiru committed:
style: run prettier on test directory

Yashiru committed:
test: improved the full flow test

alexanderattar committed:
Minor cleanup

alexanderattar committed:
Prettier
